lib/value/src/kind/collection/unknown.rs

```diff
@@ -0,0 +1,213 @@
+use std::ops::{BitAnd, BitOr, BitXor};
+
+use super::Exact;
+use crate::Kind;
+
+/// The type-state of "unknown" values in a collection.
+///
+/// That is, given a collection, it can have a set of "known" value types (e.g. we know the object
+/// collection has a field `.foo` with a type `integer`), but also a singular "unknown" value type
+/// (e.g. the array collection has an integer value at index 0, and is 3 values in size. We don't
+/// know the exact values for indices 1 and 2, but we do know that it has to be the type defined by
+/// `Unknown`).
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Unknown {
+    Any,
+    Exact(Exact),
+}
+
+impl Unknown {
+    /// Get the `any` state for `Unknown`.
+    #[must_use]
+    pub fn any() -> Self {
+        Self::Any
+    }
+
+    /// Get the `exact` state for `Unknown`.
+    #[must_use]
+    pub fn exact(exact: Exact) -> Self {
+        Self::Exact(exact)
+    }
+
+    /// Get the `json` state for `Unknown`.
+    #[must_use]
+    pub fn json() -> Self {
+        Self::exact(Exact::json())
+    }
+
+    /// Check if the state of `Unknown` is "any".
+    #[must_use]
+    pub fn is_any(&self) -> bool {
+        matches!(self, Self::Any)
+    }
+
+    /// Check if the state of `Unknown` is "exact".
+    #[must_use]
+    pub fn is_exact(&self) -> bool {
+        matches!(self, Self::Exact(_))
+    }
+
+    /// Check if `self` is a superset of `other`.
+    ///
+    /// Meaning, if `self` is `Any`, then it's always a superset of `other`, otherwise its
+    /// accumulative types need to be a superset of `other`.
+    pub fn is_superset(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Any, _) => true,
+            (Self::Exact(lhs), Unknown::Exact(rhs)) => lhs.is_superset(rhs),
+
+            // Technically, `Exact` can have all states set to `true`, which would be the same as
+            // `Unknown::Any`, but this is an invalid invariant, and considered a programming bug.
+            (Self::Exact(_), Self::Any) => false,
+        }
+    }
+}
+
+impl BitOr for Unknown {
+    type Output = Self;
+
+    /// If the state of `Unknown` fields is `any`, this method always returns `any`.
+    ///
+    /// Otherwise, a bit-or operation is performed on the `Exact` state.
+    fn bitor(self, rhs: Self) -> Self::Output {
+        match (self, rhs) {
+            (Self::Any, _) | (_, Self::Any) => Self::Any,
+            (Self::Exact(lhs), Self::Exact(rhs)) => Self::Exact(lhs | rhs),
+        }
+    }
+}
+
+impl BitXor for Unknown {
+    type Output = Self;
+
+    /// If the state of `Unknown` fields is `any`, this method always returns `any`.
```

Comment:
It should, I didn't pay much attention to these impls, other than having them for consistency, but it's actually one of the other regrets I have of the old `Kind` impl, and I've since removed these from the follow-up PR. It's just too confusing in everyday use, given the complexities that come from the conceptual model of progressive type checking.